components/script/dom/bindings/utils.rs

```diff
@@ -50,15 +52,104 @@ impl HeapSizeOf for WindowProxyHandler {
     }
 }
 
-struct ServoJSPrincipal(*mut JSPrincipals);
+//TODO make principal.rs
+pub struct ServoJSPrincipal(*mut JSPrincipals);
 
 impl ServoJSPrincipal {
     pub unsafe fn origin(&self) -> MutableOrigin {
-        let origin = getPrincipalOrigin(self.0) as *mut MutableOrigin;
+        let origin = GetPrincipalOrigin(self.0) as *mut MutableOrigin;
         (*origin).clone()
     }
 }
 
+// avadacatavra: destroy will need to retrieved the boxed origin pointer, turn it back into a box and allow it to be freed
+pub unsafe fn destroy_servo_jsprincipal(principal: &mut ServoJSPrincipal) {
+    let origin = GetPrincipalOrigin(principal.0) as *mut Box<MutableOrigin>;
```

Comment:
Casting to a `*mut Box` is incorrect; we need `*mut MutableOrigin` here so we can use [from_raw](http://doc.servo.org/std/boxed/struct.Box.html#method.from_raw) and cause its destructor to run.